### PR TITLE
Feature/multiple types

### PR DESCRIPTION
**Pull request by denis-yuen**
Upgraded youxia to handle a mixed list of instance types rather than a number of one kind of instance.

Also
- upgraded AWS library
- fixed some logging issues
- Java 8 maps are cool

### CODE REVIEW

File: youxia-deployer/src/main/java/io/cloudbindle/youxia/deployer/Deployer.java [hunk 1]
```diff
@@ -192,25 +203,80 @@ public Deployer(String[] args) {
...
+        LinkedHashMap<String, Integer> clientTypes = new LinkedHashMap<>();
+        if (options.has(totalNodesSpec)) {
+            int totalNumNodes = options.valueOf(totalNodesSpec);
+            int clientsNeeded = totalNumNodes - map.size();
+            Log.info("Need " + clientsNeeded + " more workers");
+            int clientsAfterBatching = Math.min(batchSize, clientsNeeded);
+            Log.info("After batch limit, we can requisition up to " + clientsAfterBatching + " this run");
+            clientTypes.put(defaultFlavour, clientsAfterBatching);
+        } else {
+            assert (options.has(instanceRestrictions));
```
Comment (SolomonShorser-OICR) [inline youxia-deployer/src/main/java/io/cloudbindle/youxia/deployer/Deployer.java:236]: Was the assert above only for debugging, or is it supposed to go to production?

Comment (denis-yuen) [inline youxia-deployer/src/main/java/io/cloudbindle/youxia/deployer/Deployer.java:236]: I think that it should be logically not accessible due the required unless 

```
this.instanceRestrictions = parser
                .acceptsAll(Arrays.asList("instance-types"),
                        "If specified, ansible will override the basic specified flavors (files contains a map of flavour -> # of instances)").requiredUnless(totalNodesSpec)
                .withRequiredArg().ofType(String.class);
```

File: README.md [hunk 2]
```diff
@@ -159,6 +159,15 @@ Append the parameter --openstack to deploy OpenStack instances
 
 There are also two parameters that can be used to deploy a minimum number of on-demand instances (--max-on-demand) and cap the number of on-demand instances (--min-on-demand). 
 
+It is also possible to specify a mix of instance types instead of `--total-nodes-num`. For example, the following JSON will result in 4 m1.small instances, 2 m1.xlarge, and 4 t1.micro. This should be stored in a file that is referenced using the `--instance-types` argument, as `--instance-types instance-types.json`:
+
+	{
+	"m1.small":4,
+	"m1.xlarge":2,
+	"t1.micro":4
+	}
```
Comment (SolomonShorser-OICR) [inline README.md:168]: Will this file be re-loaded every time the deployer gets a new batch of files? or would the deployer need to be restarted?

File: youxia-deployer/src/main/java/io/cloudbindle/youxia/deployer/Deployer.java [hunk 3]
```diff
@@ -603,12 +673,13 @@ private boolean runDeploymentForAWS(int clientsToDeploy, Map<String, String> ext
...
 
-    private void runDeploymentForAzure(Map<String, String> extraTags, int clientsToDeploy) throws Exception {
+    private void runDeploymentForAzure(Map<String, String> extraTags, Pair<String, Integer> clientsToDeploy) throws Exception {
         Map<String, DeploymentGetResponse> nodes = new HashMap<>();
         ComputeManagementClient azureComputeClient = ConfigTools.getAzureComputeClient();
         AzureResourceManagerClient azureResourceManagerClient = ConfigTools.getAzureResourceManagerClient();
 
-        for (int i = 0; i < clientsToDeploy; i++) {
+        // TODO: implement deployment of different types of instances on Azure
+        for (int i = 0; i < clientsToDeploy.getValue(); i++) {
```
Comment (SolomonShorser-OICR) [inline youxia-deployer/src/main/java/io/cloudbindle/youxia/deployer/Deployer.java:682]: need to worry about NullPointerException if clientsToDeploy is null??

**Pull request merged**